### PR TITLE
Issue 18855 container npe

### PR DESCRIPTION
**Pull request by wezell**
The ContainerWebAPI, which is responsible for populating what content goes into what containers, was throwing an NPE.  This is because of this line - which declared the `HttpServletRequest` as `static`.  99.9% of the time this worked, but if there were multiple threads hitting different pages, then there were cases where the request that was used to populate the container was different than the request that was viewing the page, resulting in an NPE. 

https://github.com/dotCMS/core/pull/18858/files#diff-3fcec2e785fd29108c819fbb5274c4dcL39

### CODE REVIEW

File: dotCMS/src/main/java/com/dotcms/rendering/velocity/viewtools/VelocityRequestWrapper.java [hunk 1]
```diff
@@ -1,347 +1,149 @@
...
+        }
+        return super.getRequestURI();
     }
 
-    public ServletContext getServletContext() {
-        SecurityLogger.logInfo(this.getClass(), "User trying to access ServletContext from Velocity");
-        return null;
+    @Override
+    public StringBuffer getRequestURL() {
+        return new StringBuffer(super.getScheme() + "://" + super.getServerName() + ":"
```
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotcms/rendering/velocity/viewtools/VelocityRequestWrapper.java:66]: Codacy found an issue: [Avoid concatenating nonliterals in a StringBuffer/StringBuilder constructor or append().](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5831320)

File: dotCMS/src/main/java/com/dotcms/rendering/velocity/viewtools/VelocityRequestWrapper.java [hunk 2]
```diff
@@ -1,347 +1,149 @@
...
+    }
 
-    public String getHeader ( String arg0 ) {
+    @Override
+    public String getHeader(final String header) {
 
-        if ( arg0 != null
-                && arg0.toLowerCase().equals( "user-agent" )
-                && this.getCustomUserAgentHeader() != null ) {
+        if ("user-agent".equalsIgnoreCase(header) && this.getCustomUserAgentHeader() != null) {
```
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotcms/rendering/velocity/viewtools/VelocityRequestWrapper.java:49]: Codacy found an issue: [Avoid using Literals in Conditional Statements](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5831320)

File: dotCMS/src/main/java/com/dotcms/rendering/velocity/viewtools/VelocityRequestWrapper.java [hunk 3]
```diff
@@ -1,347 +1,149 @@
...
 
-public class VelocityRequestWrapper extends javax.servlet.http.HttpServletRequestWrapper  {
+/**
+ * This class wraps the incoming HttpServletRequest to provide security and to allow uris to be
+ * overridden
+ * 
+ * @author will
+ *
+ */
+public class VelocityRequestWrapper extends javax.servlet.http.HttpServletRequestWrapper {
```
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotcms/rendering/velocity/viewtools/VelocityRequestWrapper.java:22]: Codacy found an issue: [A class which only has private constructors should be final](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5831320)

**Review by freddyucv — APPROVED** [review]

**Review by erickgonzalez — APPROVED** [review]

File: dotCMS/src/main/java/com/dotcms/rendering/velocity/viewtools/VelocityRequestWrapper.java (hunk 1 above)
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotcms/rendering/velocity/viewtools/VelocityRequestWrapper.java:66]: Codacy found an issue: [Avoid concatenating nonliterals in a StringBuffer/StringBuilder constructor or append().](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5831320)

File: dotCMS/src/main/java/com/dotcms/rendering/velocity/viewtools/VelocityRequestWrapper.java (hunk 2 above)
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotcms/rendering/velocity/viewtools/VelocityRequestWrapper.java:49]: Codacy found an issue: [Avoid using Literals in Conditional Statements](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5831320)

File: dotCMS/src/main/java/com/dotcms/rendering/velocity/viewtools/VelocityRequestWrapper.java (hunk 3 above)
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotcms/rendering/velocity/viewtools/VelocityRequestWrapper.java:22]: Codacy found an issue: [A class which only has private constructors should be final](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5831320)

File: dotCMS/src/integration-test/java/com/dotcms/rendering/velocity/viewtools/ContainerWebAPIIntegrationTest.java [hunk 4]
```diff
@@ -11,11 +11,16 @@
 import com.dotcms.contenttype.business.ContentTypeAPI;
 import com.dotcms.contenttype.model.type.BaseContentType;
 import com.dotcms.contenttype.model.type.ContentType;
-import com.dotcms.datagen.TestDataUtils;
-import com.dotcms.datagen.UserDataGen;
+import com.dotcms.datagen.*;
+import com.dotcms.enterprise.publishing.remote.handler.ContentHandler;
+import com.dotcms.enterprise.rules.RulesAPIImpl;
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotcms/rendering/velocity/viewtools/ContainerWebAPIIntegrationTest.java:16]: Codacy found an issue: [Avoid unused imports such as 'com.dotcms.enterprise.rules.RulesAPIImpl'](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5831320)

File: dotCMS/src/integration-test/java/com/dotcms/rendering/velocity/viewtools/ContainerWebAPIIntegrationTest.java [hunk 5]
```diff
@@ -11,11 +11,16 @@
 import com.dotcms.contenttype.business.ContentTypeAPI;
 import com.dotcms.contenttype.model.type.BaseContentType;
 import com.dotcms.contenttype.model.type.ContentType;
-import com.dotcms.datagen.TestDataUtils;
-import com.dotcms.datagen.UserDataGen;
+import com.dotcms.datagen.*;
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotcms/rendering/velocity/viewtools/ContainerWebAPIIntegrationTest.java:14]: Codacy found an issue: [Avoid unused imports such as 'com.dotcms.datagen'](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5831320)

File: dotCMS/src/main/java/com/dotcms/rendering/velocity/viewtools/VelocityRequestWrapper.java (hunk 2 above)
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotcms/rendering/velocity/viewtools/VelocityRequestWrapper.java:49]: Codacy found an issue: [Avoid using Literals in Conditional Statements](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5831320)

File: dotCMS/src/integration-test/java/com/dotcms/rendering/velocity/viewtools/ContainerWebAPIIntegrationTest.java [hunk 6]
```diff
@@ -220,4 +239,131 @@ private Container createContainer(final User adminUser) throws DotDataException,
...
+        final Host host = new SiteDataGen().nextPersisted();
+        final Template template = new TemplateDataGen().nextPersisted();
+        final HTMLPageAsset htmlPageAsset = new HTMLPageDataGen(host, template).nextPersisted();
+        final Container container = new ContainerDataGen().nextPersisted();
+
+        final HttpServletRequest request = mockHttpServletRequest(mode);
+
+        final HttpServletResponse response = mock(HttpServletResponse.class);
+
+        final MultiTree multiTree = new MultiTreeDataGen()
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotcms/rendering/velocity/viewtools/ContainerWebAPIIntegrationTest.java:266]: Codacy found an issue: [Avoid unused local variables such as 'multiTree'.](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5831320)

File: dotCMS/src/integration-test/java/com/dotcms/rendering/velocity/viewtools/ContainerWebAPIIntegrationTest.java [hunk 7]
```diff
@@ -25,15 +30,29 @@
 import com.dotmarketing.exception.DotDataException;
 import com.dotmarketing.exception.DotSecurityException;
 import com.dotmarketing.portlets.containers.model.Container;
+import com.dotmarketing.portlets.contentlet.model.Contentlet;
+import com.dotmarketing.portlets.htmlpageasset.model.HTMLPageAsset;
+import com.dotmarketing.portlets.personas.model.Persona;
+import com.dotmarketing.portlets.rules.model.Condition;
+import com.dotmarketing.portlets.rules.model.ConditionGroup;
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotcms/rendering/velocity/viewtools/ContainerWebAPIIntegrationTest.java:37]: Codacy found an issue: [Avoid unused imports such as 'com.dotmarketing.portlets.rules.model.ConditionGroup'](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5831320)

File: dotCMS/src/integration-test/java/com/dotcms/rendering/velocity/viewtools/ContainerWebAPIIntegrationTest.java [hunk 8]
```diff
@@ -11,11 +11,16 @@
 import com.dotcms.contenttype.business.ContentTypeAPI;
 import com.dotcms.contenttype.model.type.BaseContentType;
 import com.dotcms.contenttype.model.type.ContentType;
-import com.dotcms.datagen.TestDataUtils;
-import com.dotcms.datagen.UserDataGen;
+import com.dotcms.datagen.*;
+import com.dotcms.enterprise.publishing.remote.handler.ContentHandler;
+import com.dotcms.enterprise.rules.RulesAPIImpl;
+import com.dotcms.rendering.velocity.directive.DotParse;
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotcms/rendering/velocity/viewtools/ContainerWebAPIIntegrationTest.java:17]: Codacy found an issue: [Avoid unused imports such as 'com.dotcms.rendering.velocity.directive.DotParse'](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5831320)

File: dotCMS/src/integration-test/java/com/dotcms/rendering/velocity/viewtools/ContainerWebAPIIntegrationTest.java [hunk 9]
```diff
@@ -25,15 +30,29 @@
 import com.dotmarketing.exception.DotDataException;
 import com.dotmarketing.exception.DotSecurityException;
 import com.dotmarketing.portlets.containers.model.Container;
+import com.dotmarketing.portlets.contentlet.model.Contentlet;
+import com.dotmarketing.portlets.htmlpageasset.model.HTMLPageAsset;
+import com.dotmarketing.portlets.personas.model.Persona;
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotcms/rendering/velocity/viewtools/ContainerWebAPIIntegrationTest.java:35]: Codacy found an issue: [Avoid unused imports such as 'com.dotmarketing.portlets.personas.model.Persona'](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5831320)

File: dotCMS/src/main/java/com/dotcms/rendering/velocity/viewtools/VelocityRequestWrapper.java (hunk 3 above)
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotcms/rendering/velocity/viewtools/VelocityRequestWrapper.java:22]: Codacy found an issue: [A class which only has private constructors should be final](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5831320)

File: dotCMS/src/integration-test/java/com/dotcms/rendering/velocity/viewtools/ContainerWebAPIIntegrationTest.java [hunk 10]
```diff
@@ -25,15 +30,29 @@
...
+import com.dotmarketing.portlets.rules.model.Condition;
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotcms/rendering/velocity/viewtools/ContainerWebAPIIntegrationTest.java:36]: Codacy found an issue: [Avoid unused imports such as 'com.dotmarketing.portlets.rules.model.Condition'](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5831320)

File: dotCMS/src/integration-test/java/com/dotcms/rendering/velocity/viewtools/ContainerWebAPIIntegrationTest.java [hunk 11]
```diff
@@ -11,11 +11,16 @@
 import com.dotcms.contenttype.business.ContentTypeAPI;
 import com.dotcms.contenttype.model.type.BaseContentType;
 import com.dotcms.contenttype.model.type.ContentType;
-import com.dotcms.datagen.TestDataUtils;
-import com.dotcms.datagen.UserDataGen;
+import com.dotcms.datagen.*;
+import com.dotcms.enterprise.publishing.remote.handler.ContentHandler;
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotcms/rendering/velocity/viewtools/ContainerWebAPIIntegrationTest.java:15]: Codacy found an issue: [Avoid unused imports such as 'com.dotcms.enterprise.publishing.remote.handler.ContentHandler'](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5831320)

File: dotCMS/src/main/java/com/dotcms/rendering/velocity/viewtools/VelocityRequestWrapper.java (hunk 1 above)
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotcms/rendering/velocity/viewtools/VelocityRequestWrapper.java:66]: Codacy found an issue: [Avoid concatenating nonliterals in a StringBuffer/StringBuilder constructor or append().](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5831320)

File: dotCMS/src/integration-test/java/com/dotcms/rendering/velocity/viewtools/ContainerWebAPIIntegrationTest.java [hunk 12]
```diff
@@ -25,15 +30,29 @@
...
 import com.liferay.portal.model.User;
 import com.liferay.portal.util.WebKeys;
 import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.List;
 import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
 import javax.servlet.http.HttpSession;
+
+import com.liferay.util.StringPool;
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotcms/rendering/velocity/viewtools/ContainerWebAPIIntegrationTest.java:51]: Codacy found an issue: [Avoid unused imports such as 'com.liferay.util.StringPool'](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5831320)

File: dotCMS/src/integration-test/java/com/dotcms/rendering/velocity/viewtools/ContainerWebAPIIntegrationTest.java (hunk 7 above)
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotcms/rendering/velocity/viewtools/ContainerWebAPIIntegrationTest.java:37]: Codacy found an issue: [Avoid unused imports such as 'com.dotmarketing.portlets.rules.model.ConditionGroup'](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5831320)

File: dotCMS/src/integration-test/java/com/dotcms/rendering/velocity/viewtools/ContainerWebAPIIntegrationTest.java (hunk 4 above)
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotcms/rendering/velocity/viewtools/ContainerWebAPIIntegrationTest.java:16]: Codacy found an issue: [Avoid unused imports such as 'com.dotcms.enterprise.rules.RulesAPIImpl'](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5831320)

File: dotCMS/src/integration-test/java/com/dotcms/rendering/velocity/viewtools/ContainerWebAPIIntegrationTest.java (hunk 6 above)
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotcms/rendering/velocity/viewtools/ContainerWebAPIIntegrationTest.java:266]: Codacy found an issue: [Avoid unused local variables such as 'multiTree'.](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5831320)

File: dotCMS/src/main/java/com/dotcms/rendering/velocity/viewtools/VelocityRequestWrapper.java (hunk 2 above)
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotcms/rendering/velocity/viewtools/VelocityRequestWrapper.java:49]: Codacy found an issue: [Avoid using Literals in Conditional Statements](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5831320)

File: dotCMS/src/integration-test/java/com/dotcms/rendering/velocity/viewtools/ContainerWebAPIIntegrationTest.java (hunk 10 above)
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotcms/rendering/velocity/viewtools/ContainerWebAPIIntegrationTest.java:36]: Codacy found an issue: [Avoid unused imports such as 'com.dotmarketing.portlets.rules.model.Condition'](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5831320)

File: dotCMS/src/integration-test/java/com/dotcms/rendering/velocity/viewtools/ContainerWebAPIIntegrationTest.java (hunk 11 above)
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotcms/rendering/velocity/viewtools/ContainerWebAPIIntegrationTest.java:15]: Codacy found an issue: [Avoid unused imports such as 'com.dotcms.enterprise.publishing.remote.handler.ContentHandler'](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5831320)

File: dotCMS/src/integration-test/java/com/dotcms/rendering/velocity/viewtools/ContainerWebAPIIntegrationTest.java (hunk 8 above)
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotcms/rendering/velocity/viewtools/ContainerWebAPIIntegrationTest.java:17]: Codacy found an issue: [Avoid unused imports such as 'com.dotcms.rendering.velocity.directive.DotParse'](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5831320)

File: dotCMS/src/integration-test/java/com/dotcms/rendering/velocity/viewtools/ContainerWebAPIIntegrationTest.java (hunk 12 above)
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotcms/rendering/velocity/viewtools/ContainerWebAPIIntegrationTest.java:51]: Codacy found an issue: [Avoid unused imports such as 'com.liferay.util.StringPool'](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5831320)

File: dotCMS/src/main/java/com/dotcms/rendering/velocity/viewtools/VelocityRequestWrapper.java (hunk 3 above)
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotcms/rendering/velocity/viewtools/VelocityRequestWrapper.java:22]: Codacy found an issue: [A class which only has private constructors should be final](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5831320)

File: dotCMS/src/integration-test/java/com/dotcms/rendering/velocity/viewtools/ContainerWebAPIIntegrationTest.java (hunk 9 above)
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotcms/rendering/velocity/viewtools/ContainerWebAPIIntegrationTest.java:35]: Codacy found an issue: [Avoid unused imports such as 'com.dotmarketing.portlets.personas.model.Persona'](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5831320)

File: dotCMS/src/integration-test/java/com/dotcms/rendering/velocity/viewtools/ContainerWebAPIIntegrationTest.java (hunk 5 above)
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotcms/rendering/velocity/viewtools/ContainerWebAPIIntegrationTest.java:14]: Codacy found an issue: [Avoid unused imports such as 'com.dotcms.datagen'](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5831320)

File: dotCMS/src/main/java/com/dotcms/rendering/velocity/viewtools/VelocityRequestWrapper.java (hunk 1 above)
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotcms/rendering/velocity/viewtools/VelocityRequestWrapper.java:66]: Codacy found an issue: [Avoid concatenating nonliterals in a StringBuffer/StringBuilder constructor or append().](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5831320)

Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotcms/rendering/velocity/viewtools/VelocityRequestWrapper.java:66]: Codacy found an issue: [Avoid concatenating nonliterals in a StringBuffer/StringBuilder constructor or append().](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5831320)

File: dotCMS/src/main/java/com/dotcms/rendering/velocity/viewtools/VelocityRequestWrapper.java (hunk 2 above)
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotcms/rendering/velocity/viewtools/VelocityRequestWrapper.java:49]: Codacy found an issue: [Avoid using Literals in Conditional Statements](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5831320)

File: dotCMS/src/main/java/com/dotcms/rendering/velocity/viewtools/VelocityRequestWrapper.java (hunk 3 above)
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotcms/rendering/velocity/viewtools/VelocityRequestWrapper.java:22]: Codacy found an issue: [A class which only has private constructors should be final](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5831320)

File: dotCMS/src/integration-test/java/com/dotcms/rendering/velocity/viewtools/ContainerWebAPIIntegrationTest.java [hunk 13]
```diff
@@ -220,4 +240,131 @@ private Container createContainer(final User adminUser) throws DotDataException,
...
+        final Host host = new SiteDataGen().nextPersisted();
+        final Template template = new TemplateDataGen().nextPersisted();
+        final HTMLPageAsset htmlPageAsset = new HTMLPageDataGen(host, template).nextPersisted();
+        final Container container = new ContainerDataGen().nextPersisted();
+
+        final HttpServletRequest request = mockHttpServletRequest(mode);
+
+        final HttpServletResponse response = mock(HttpServletResponse.class);
+
+        final MultiTree multiTree = new MultiTreeDataGen()
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotcms/rendering/velocity/viewtools/ContainerWebAPIIntegrationTest.java:267]: Codacy found an issue: [Avoid unused local variables such as 'multiTree'.](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5831320)

**Pull request merged**